sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala

```diff
@@ -128,6 +129,93 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
       TableIdentifier("tempTable"), ignoreIfNotExists = true, purge = false)
   }
 
+  test("SPARK-21079 - analyze table with location different than that of individual partitions") {
+    def queryTotalSize(tableName: String): BigInt =
+      spark.table(tableName).queryExecution.analyzed.stats(conf).sizeInBytes
+
+    val tableName = "analyzeTable_part"
+    withTable(tableName) {
+      // Create 4 paths: one to use as table location and one for each of the 3 partitions
+      withTempPaths(numPaths = 4) {
+        case tablePath :: partitionPaths =>
+          sql(
+            s"""
+               |CREATE TABLE ${tableName} (key STRING, value STRING) PARTITIONED BY (ds STRING)
+               |LOCATION '${tablePath}'
+             """.stripMargin).collect()
+
+          val partitionDates = List("2010-01-01", "2010-01-02", "2010-01-03")
+          partitionDates.zip(partitionPaths).foreach {
+            case (ds, path) =>
+              sql(
+                s"""
+                   |ALTER TABLE ${tableName} ADD PARTITION (ds='${ds}')
+                   |LOCATION '${path.toString}'
+                """.stripMargin).collect()
+              sql(
+                s"""
+                   |INSERT INTO TABLE ${tableName} PARTITION (ds='${ds}')
+                   |SELECT * FROM src
+                """.stripMargin).collect()
+          }
+
+          sql(s"ANALYZE TABLE ${tableName} COMPUTE STATISTICS noscan")
+
+          assert(queryTotalSize(tableName) === BigInt(17436))
+      }
+    }
+  }
+
+  test("SPARK-21079 - analyze partitioned table with only a subset of partitions visible") {
+    def queryTotalSize(tableName: String): BigInt =
+      spark.table(tableName).queryExecution.analyzed.stats(conf).sizeInBytes
+
+    val sourceTableName = "analyzeTable_part"
+    val tableName = "analyzeTable_part_vis"
+    withTable(sourceTableName, tableName) {
+      withTempPath {
+        path =>
+          // Create a table with 3 partitions all located under a single top-level directory 'path'
+          sql(
+            s"""
+               |CREATE TABLE ${sourceTableName} (key STRING, value STRING)
+               |PARTITIONED BY (ds STRING)
+               |LOCATION '${path}'
+             """.stripMargin).collect()
+
+          val partitionDates = List("2010-01-01", "2010-01-02", "2010-01-03")
+          partitionDates.foreach {
+            ds =>
+              sql(
+                s"""
+                   |INSERT INTO TABLE ${sourceTableName} PARTITION (ds='${ds}')
```

Comment:
`${sourceTableName}` -> `$sourceTableName`
`${ds}` -> `$ds`

Please fix the other similar issues